src/adaptors/tcp/tcp_adaptor.c

```diff
@@ -1369,9 +1433,39 @@ static bool qdr_tcp_create_egress_connection(qd_tcp_connector_t      *connector,
     tc->is_egress_dispatcher_conn = false;
 
     tc->require_tls = !!connector->tls_domain;
+    qd_message_t *msg = qdr_delivery_message(initial_delivery);
     if (tc->require_tls) {
         // Since TLS is required on this connection, try to initialize TLS attributes
         // from the associated sslProfile.
+        char *alpn_protocol;
+        set_alpn_protocol(tc, msg, &alpn_protocol);
```

Comment:
Something to consider: walking the message's application properties map is _very_ expensive (remember old router message annotations?).

The problem here is that we're doing it *twice* this this change.

See qdr_associate_vflow_flows() - it is done there as well.

I would recommend creating a utility routine that given the message's QD_FIELD_APPLICATION_PROPERTIES extracts & processes all the interesting stuff *in one pass*.  It looks like (currently) all that is needed is the vflow id and the alpn protocol (if tls is in use) for now.

Reply:
Done !